tests/integration.rs

```diff
@@ -0,0 +1,433 @@
+    // time that Pebble starts up. This is the issuer that signs the randomly generated
+    // intermediate certificate returned as part of ACME issued certificate chains.
+    async fn issuer_roots(&self) -> Result<RootCertStore, Box<dyn StdError>> {
+        let client = self.http_client();
+
+        let req = Request::builder()
+            .method(Method::GET)
+            .uri(format!("{}/roots/0", self.pebble_management_url()))
+            .header(CONTENT_TYPE, "application/json")
+            .body(Full::default())?;
+
+        let resp = client.request(req).await?;
+        if resp.status() != 200 {
+            return Err(format!("unexpected /roots/0 response status: {}", resp.status()).into());
+        }
+
+        let body = resp.collect().await?.aggregate();
+        let mut pem = String::new();
+        body.reader().read_to_string(&mut pem)?;
+        let root = CertificateDer::from_pem_slice(pem.as_bytes())?;
+
+        let mut roots = RootCertStore::empty();
+        roots.add_parsable_certificates(vec![root]);
+        assert_eq!(roots.len(), 1);
+        Ok(roots)
+    }
+
+    fn challenge_management_url(&self) -> &str {
+        "http://127.0.0.1:8055" // Default.
+    }
+
+    fn pebble_management_url(&self) -> String {
+        format!("https://{}", &self.config.pebble.management_listen_address)
+    }
+
+    fn directory_url(&self) -> String {
+        format!("https://{}/dir", &self.config.pebble.listen_address)
+    }
+}
+
+#[derive(Clone, Serialize)]
+struct Config {
+    pebble: PebbleConfig,
+}
+
+#[derive(Clone, Serialize)]
+struct PebbleConfig {
+    #[serde(rename = "listenAddress")]
+    listen_address: String,
+    #[serde(rename = "managementListenAddress")]
+    management_listen_address: String,
+    certificate: String,
+    #[serde(rename = "privateKey")]
+    private_key: String,
+    #[serde(rename = "httpPort")]
+    http_port: u16,
+    #[serde(rename = "tlsPort")]
+    tls_port: u16,
+    #[serde(rename = "ocspResponderURL")]
+    ocsp_responder_url: String,
+    #[serde(rename = "externalAccountBindingRequired")]
+    external_account_binding_required: bool,
+    #[serde(rename = "domainBlocklist")]
+    domain_blocklist: Vec<String>,
+    retry_after: RetryConfig,
+    profiles: HashMap<String, Profile>,
+}
+
+#[derive(Clone, Serialize)]
+struct RetryConfig {
+    authz: u32,
+    order: u32,
+}
+
+#[derive(Clone, Serialize)]
+struct Profile {
+    description: String,
+    #[serde(rename = "validityPeriod")]
+    validity_period: u32,
+}
+
+struct Subprocess(Option<Child>);
+
+impl Subprocess {
+    fn new(cmd: &mut Command) -> io::Result<Self> {
+        Ok(Self(Some(cmd.spawn()?)))
+    }
+}
+
+impl Drop for Subprocess {
+    fn drop(&mut self) {
+        if let Some(mut child) = self.0.take() {
+            child.kill().expect("failed to kill subprocess");
+            child.wait().expect("failed to wait for killed subprocess");
+        }
+    }
+}
```

Comment:
Why do we need these shenanigans? Add some comments?

Reply:
Can you be more specific? I'm not sure what part you think is shenanigans :laughing: The kill and wait? the option take? doing it in drop?

Reply:
The whole wrapping of the child process and the `Drop` impl.

Reply:
I added some comments. If we spawn the `Child` processes but don't explicitly `kill()` and `wait()` they'll be left running after the test ends. We could do this work manually at the end of `main` but it feels more Rust-y to have this all handled consistently for both sub-processes through a wrapper type and a `Drop` impl. The inspiration comes from [rustls-openssl-compat](https://github.com/cpu/rustls-openssl-compat/blob/5247252014aa62079708b7a9c9095a18d6e3f659/tests/runner.rs#L730-L765) where it does a bit more work w.r.t capturing output.